src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        private Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> insertNode;
+        if (size == 0) {
+            head = tail = new Node<>(null, value, null);
+            size++;
+            return;
+        }
+        insertNode = new Node<>(tail, value, null);
+        tail.next = insertNode;
+        tail = insertNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index: " + index + "out of bounds, Size: " + size);
+        }
+        if (isEmpty() || index == size) {
+            add(value);
+            return;
+        }
+        Node<T> insertNode;
+        Node<T> node = getNode(index);
+        if (index == 0) {
+            insertNode = new Node<>(null, value, head);
+            head = insertNode;
+            head.prev = insertNode;
+            size++;
+            return;
+        }
+        insertNode = new Node<>(node.prev, value, node);
+        node.prev.next = insertNode;
+        node.prev = insertNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexException(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexException(index);
+        Node<T> node = getNode(index);
+        T changedItem = node.item;
+        node.item = value;
+        return changedItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexException(index);
+        Node<T> node = getNode(index);
+        return unlink(node);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (node.item == object || (node.item != null && node.item.equals(object))) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
```

Comment:
we can call `checkIndex` method into `getNode(int index)` method, this will reduce the number of duplicated code